### PR TITLE
Apply BZ 65358: Improve EL matching of varargs methods. 

### LINKED ISSUE #17444: Pull in BZ 65358  -- Varargs Method Matching  (EL Patch)
Description:

It was reported to us that EL may select the wrong method after we applied a previous EL patch in 21.0.0.3. 

We created an issue in the Tomcat EL community here: 
https://bz.apache.org/bugzilla/show_bug.cgi?id=65358

And they provided patches which we are pulling in. 

The complete fix was done in two parts in Tomcat 10.0.7 & 10.0.8. 

I pulled them in with two PRs: #17447 and #17896 for **EL-3.0.**

______
Once we upgrade to EL 10.0.8 or higher for expressionLanguage-4.0 [beta feature, currently], this varargs issue should be fully completed (#18322)

### PR DESCRIPTION
**Pull request by volosied**
fixes #17444

See the original patch here: https://github.com/apache/tomcat/commit/4f5a35563f83e28aa341f69c9dd7a856a0aed895#

In a nutshell, it keeps track of the varargs when selecting the best method. Method with the fewest varargs would then selected. 

Fix applies to EL-3.0. 

This is also fixed for expressionLanague-4.0 in PR  #17669 (Update EL to 10.0.7)

### CODE REVIEW

File: dev/com.ibm.ws.org.apache.jasper.el/src/org/apache/el/util/ReflectionUtil.java [hunk 1]
```diff
@@ -178,33 +179,58 @@ public static Method getMethod(EvaluationContext ctx, Object base, Object proper
...
+            }
+            if (m.isVarArgs() && paramCount > mParamCount && paramValues != null &&
+                    paramValues.length != paramCount) {
+                // Might match a different varargs method
+                continue;
+            }
+            if (!m.isVarArgs() && paramValues != null && paramCount != paramValues.length) {
+                // Might match a different varargs method
+                continue;
+            }
```
Comment (volosied) [inline dev/com.ibm.ws.org.apache.jasper.el/src/org/apache/el/util/ReflectionUtil.java:203]: Per the comment in the method description, this method was out of sync with the one in Util.java.  I added the relevant portions this method to have it matching again.

File: dev/com.ibm.ws.org.apache.jasper.el/src/org/apache/el/util/ReflectionUtil.java [hunk 2]
```diff
@@ -508,16 +543,43 @@ public int compareTo(MatchResult o) {
...
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + assignable;
+            result = prime * result + (bridge ? 1231 : 1237);
+            result = prime * result + coercible;
+            result = prime * result + exact;
+            result = prime * result + varArgs;
+            return result;
+        }
```
Comment (volosied) [inline dev/com.ibm.ws.org.apache.jasper.el/src/org/apache/el/util/ReflectionUtil.java:581]: Same as before for MatchResult. Keep Util.java and RelfectionUtil.java in sync.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.el.3.0_fat,com.ibm.ws.jsf.2.2_fat

**Review by epj — CHANGES_REQUESTED** [review]
The Tomcat fix also includes tests.  The tests (or equivalents that fit our harness) should be added to our FAT in this PR also.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.el.3.0_fat,com.ibm.ws.jsf.2.2_fat

**Review by epj — APPROVED** [review]

**Review by wtlucy — APPROVED** [review]

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsp.2.3_fat,com.ibm.ws.el.3.0_fat,com.ibm.ws.jsf.2.2_fat

**Comment by volosied** [comment]
#run-libby-bot

**Comment by volosied** [comment]
I labelled the issue as release bug, but I think this PR should also have it.

**Pull request merged**